aiohttp/web_protocol.py

```diff
@@ -284,7 +286,8 @@ async def shutdown(self, timeout: Optional[float] = 15.0) -> None:
             self._keepalive_handle.cancel()
 
         # Wait for graceful handler completion
-        if self._handler_waiter is not None:
+        if self._request_in_progress:
+            self._handler_waiter = self._loop.create_future()
```

Comment:
If this is genuinely a notable performance improvement, I think a comment should be added to explain this. Otherwise, it looks like code ready to be refactored to something simpler.

Reply:
Will do. I’ll add a profile for this one as well. Any time we can get rid of creating a future it’s always a nice performance bump.

Reply:
Added comments, and profile above.